docs/sql/statements/grant.rst

```diff
@@ -18,17 +19,24 @@ Synopsis
 
   GRANT { { DQL | DML | DDL | AL [,...] } | ALL [ PRIVILEGES ] }
   [ON {SCHEMA | TABLE | VIEW} identifier [, ...]]
-  TO user_name [, ...];
+  TO name [, ...]
+
+  GRANT role_name_to_grant [, ...] TO name [, ...]
 
 Description
 ===========
 
-``GRANT`` is a management statement to grant one or many privileges
-on the whole cluster or on a specific object to one or many existing users.
+``GRANT`` is a management statement which comes in two flavours.
 
+The first one is used to grant one or many privileges on the whole cluster or on
+a specific object to one or many existing users or roles.
 ``ON {SCHEMA | TABLE | VIEW}`` is optional, if not specified the privilege will
 be granted on the ``CLUSTER`` level.
 
+With the second one, ``GRANT`` can be used to grant one or more roles to one or
+many others or roles. Thus, the users (or roles) inherit the privileges of the
+roles which they are :ref:`granted<roles_inheritance>`.
```

Comment:
Hm, that's actually depends on the order. 
Denying a privilege which user doesn't have is a no-op

So: 
grant DQL -> DENY DQL = no DQL
DENY DQL -> no-op -> GRANT DQL = has DQL

I think we need to agree on the order (ie which sequence is preferable), document it and make sure privileges resolution happens in that order (first deny or first grant).

Or I'm missing something?

Reply:
I am thinking of a situation where a role may grant DQL on a schema and another role deny on a specific table for instance, I would expect access to be denied in this case regardless of the order in which the permissions were configured.

Reply:
NB: `DENY` <> `REVOKE`

Reply:
>NB: DENY <> REVOKE

Oh yes, I was writing this example with REVOKE in mind. (so still not clear to me hw to deal with GRANT, REVOKE order)

On DENY - it's not supported in PG, couldn't find in SQL standard but in MSSQL DENY has precedence over GRANT.
https://learn.microsoft.com/en-us/sql/t-sql/statements/deny-transact-sql?view=sql-server-ver16
> DENY takes precedence over all permissions

Reply:
@BaurzhanSakhariev it's about DENY a priv like DQL but the order of resolution from a user to its parents. So, 
1. if we have `user`->`role_a`, and `user` has a `deny DQL`, and `role_a` has a `grant dql`, `deny` should be respected, no question (I guess). 
2. If we have `user`->`role_a`->`role_b`, `user` has nothing, `role_a` has `deny DQL`, `role_b` has `grant DQL` then `deny` should be respected.
3. If we have `user`->`role_a`->`role_b`, `user` has `grant dql`, `role_a` has `deny DQL`, `role_b` has `grant DQL` then `grant` should be respected as it's directly on the user.
4. If we have `user`->`role_a`, `role_b` (same level) `user` has  nothing, `role_a` has `deny DQL`, `role_b` has `grant DQL` then? currently we don't keep order of role grants to a user, and I think it's fine to keep it that way and say that this is ambiguous and it's up to the user to not end up like that.

Reply:
we can continue the discussion here, but I'd suggest to merge this docs PR, and update this section of the docs once we resolve it, in another PR since it requires code changes.

Reply:
An example of 4 could be a case where we have a `finance_department` role with permissions on the `finance` schema and an `interns` role to which we want to `DENY DML ON SCHEMA finance`. There will be interns working in finance and they should not have modify permissions, interns may also have access to other data the `finance_department` would not have access to, so they are 2 separate groups, not a hierarchy, but some interns will work in finance.

Reply:
Thx @hlcianfagna, sounds like a valid use case. 
@mfussenegger @seut could you also share your opinion?

Reply:
After discussion with @seut, we agreed to implement the full DENY precedence:
https://github.com/crate/crate/pull/15278